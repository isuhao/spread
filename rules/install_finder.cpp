#include "install_finder.hpp"
#include <boost/filesystem.hpp>
#include <set>

using namespace Spread;

InstallFinder::InstallFinder(const RuleFinder &_rules, Cache::CacheIndex &_cache)
  : rules(_rules), cache(_cache) {}

void InstallFinder::handleDeps(const DepList &deps, ActionMap &output)
{
  for(int i=0; i<deps.size(); i++)
    {
      const std::string &dest = deps[i].first;
      const Hash &hash = deps[i].second;

      // Check if this hash has already been resolved.
      ActionMap::iterator it = output.find(hash);
      if(it != output.end())
        {
          // Add ourself to it's destination list, if we have one
          it->second.addDest(dest);

          continue;
        }

      // Hash not previously resolved. Check if the cache knows
      // anything about the hash.
      int stat = Cache::CI_ElseWhere;
      if(dest != "")
        stat = cache.getStatus(dest, hash);

      if(stat == Cache::CI_Match)
        {
          /* Add copy action with empty destination. This allows other
             tasks to look up the same action later and add multiple
             destinations.

             If no other actions need this file, then the action is
             ignored, which is OK since the file is already installed.
          */
          output[hash] = Action(dest);
          continue;
        }

      if(stat == Cache::CI_ElseWhere)
        {
          // File either exists elsewhere, or dest == "". Get cache
          // location, if any.
          const std::string &existing = cache.findHash(hash);

          if(existing != "")
            {
              // Yup, it exists. Add a copy operation.
              output[hash] = Action(existing, dest);
              continue;
            }
        }

      /* If we are here, no existing file met our dependency. Look
         up the ruleset to see how else we can fetch the file.
      */
      const Rule *r = rules.findRule(hash);

      if(r)
        {
          bool found = false;

          // Set up the output action
          Action a(r);

          // Loop through and add the action for all the hashes it
          // produces, not just the one we are looking for.
          for(int k=0; k<r->outputs.size(); k++)
            {
              const Hash &h = r->outputs[k];
              output[h] = a;
              if(h == hash)
                {
                  // Add our output destination to the right output
                  output[h].addDest(dest);
                  found = true;
                }
            }
          assert(found);

          // Then loop through this rule's dependencies, and expand
          // those as well
          DepList subdeps;
          for(int k=0; k<r->deps.size(); k++)
            {
              const Hash &h = r->deps[k];
              subdeps.push_back(DepPair("",h));
              handleDeps(subdeps, output);
            }

          /* TODO: Handle recursion. Recursion happens when a hash
             implicitly depends upon itself. We would also have to
             figure out what to do in the case of recursion. At the
             moment though, we just assume our ruleset is acyclic.
          */
          continue;
        }

      // There was no way to resolve this hash. An empty copy action
      // represents an unhandled dependency.
      output[hash] = Action("", dest);
    }
}
