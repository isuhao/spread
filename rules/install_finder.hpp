#ifndef __SPREAD_INSTALL_FINDER_HPP_
#define __SPREAD_INSTALL_FINDER_HPP_

#include "cache/index.hpp"
#include "hash/hash.hpp"
#include "rulefinder.hpp"
#include "actions.hpp"
#include <vector>

namespace Spread
{
  struct InstallFinder
  {
    InstallFinder(const RuleFinder &rules, Cache::CacheIndex &cache);

    /* Add a dependency. A dependency is a request for a specific
       object 'hash' to be installed in location 'file'.

       Several dependencies may overwrite the same location. It's
       always the one added last that takes effect. This can be used
       intentionally to install eg. a game, and then a patch on top of
       it.
     */
    void addDep(const std::string &file, const Hash &hash)
    { deps.push_back(DepPair(file,hash)); }

    /* Add a "blind" archive dependency. This means that we do not
       know the exact content of an archive at this point, we just
       want it unpacked into the given location.

       Using this saves you the trouble of keeping and distributing
       directory files for all the archives.
     */
    void addBlind(const std::string &path, const Hash &arcHash)
    { blinds.push_back(DepPair(path, arcHash)); }

    /* Perform the rule resolution. Takes the list of input
       dependencies in 'deps' and converts it to a list of actions in
       'output'.

       Returns true on success, or false if the output contains unmet
       dependencies.
     */
    bool perform(ActionMap &output) { return handleDeps(deps, output, true); }

    typedef std::pair<std::string,Hash> DepPair;
    typedef std::vector<DepPair> DepList;

    /* This is the struct's input data. You can set it up through the
       addDep/addBlind functions above, or manually if you prefer.
       Then run perform() to produce the output.

       Running perform() does not change the depslists, so the struct
       is reusable.
     */
    DepList deps, blinds;

  private:
    const RuleFinder &rules;
    Cache::CacheIndex &cache;
    bool handleDeps(const DepList &deps, ActionMap &output, bool baseLevel=false);
  };
}
#endif
