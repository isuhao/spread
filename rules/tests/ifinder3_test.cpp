#include <iostream>
#include "install_finder.hpp"

using namespace std;
using namespace Spread;

Hash zipfile("zipfile", 7);

struct MyFinder : RuleFinder
{
  Rule ziprule;
  Rule zipurl;

  MyFinder()
  {
    ziprule.addDep(zipfile);
    ziprule.ruleString = "ZIP rule";

    zipurl.addOut(zipfile);
    zipurl.ruleString = "URL for zipfile";
  }

  const Rule *findRule(const Hash &hash) const
  {
    if(hash == zipfile)
      return &zipurl;

    return NULL;
  }

  // Not used in this test
  void findAllRules(const Hash&, RuleList&) const { assert(0); }
};

MyFinder rules;
Cache::CacheIndex cache;

void print(InstallFinder &inf)
{
  ActionMap output;
  bool isOk = inf.perform(output);

  cout << "\nHandled " << inf.deps.size() << " input dependencies into "
       << output.size() << " hashes (OK=" << isOk << "):\n";
  ActionMap::iterator it;
  for(it = output.begin(); it != output.end(); it++)
    {
      Action &a = it->second;

      cout << it->first << " : ";

      if(a.isNone())
        {
          cout << "UNHANDLED\n";
        }
      else if(a.isRule())
        {
          const Rule &r = *a.rule;

          cout << r.ruleString << endl;

          for(int i=0; i<r.deps.size(); i++)
            cout << "  <= " << r.deps[i] << endl;
        }
      else if(a.isCopy())
        {
          cout << "COPY from " << a.source << endl;
        }
      else assert(0);

      std::set<std::string>::iterator it2;
      for(it2 = a.destlist.begin(); it2 != a.destlist.end(); it2++)
        cout << "  => " << *it2 << endl;
    }
}

int main()
{
  InstallFinder inf(rules, cache);
  inf.addBlind("some/destination/", zipfile);
  print(inf);

  return 0;
}
