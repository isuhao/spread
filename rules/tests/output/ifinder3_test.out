
Handled 0 input dependencies into 2 hashes (OK=1):
ARC_nOYm : Blind Archive Unpack Rule
  <= nOYmHbSVYgh-BC4bhVen8JWujb124ib0UNBaY1UjMXkH
  => /home/mortennk/koding/spread/rules/tests/some/destination/
nOYmHbSVYgh-BC4bhVen8JWujb124ib0UNBaY1UjMXkH : URL for zipfile
